File: bindings/java-jni/src/main/java/org/maplibre/nativejni/map/NorthOrientation.java
package org.maplibre.nativejni.map;

/** North orientation used by map viewport options. */
public final class NorthOrientation {
  public static final NorthOrientation UP = new NorthOrientation(0, "UP");
  public static final NorthOrientation RIGHT = new NorthOrientation(1, "RIGHT");
  public static final NorthOrientation DOWN = new NorthOrientation(2, "DOWN");
  public static final NorthOrientation LEFT = new NorthOrientation(3, "LEFT");

  private final int nativeValue;
  private final String name;

  private NorthOrientation(int nativeValue, String name) {
    this.nativeValue = nativeValue;
    this.name = name;
  }

  public int nativeValue() {
    if (name == null) {
      throw new IllegalArgumentException("Unknown north orientation cannot be used as an input");
    }
    return nativeValue;
  }

  public int rawValue() {
    return nativeValue;
  }

  public static NorthOrientation fromNative(int nativeValue) {
    return switch (nativeValue) {
      case 0 -> UP;
      case 1 -> RIGHT;
      case 2 -> DOWN;
      case 3 -> LEFT;
      default -> new NorthOrientation(nativeValue, null);
    };
  }

  @Override
  public boolean equals(Object other) {
    return other instanceof NorthOrientation value && nativeValue == value.nativeValue;
  }

  @Override
  public int hashCode() {
    return Integer.hashCode(nativeValue);
  }

  @Override
  public String toString() {
    return name != null ? name : "NorthOrientation(" + nativeValue + ")";
  }
}

File: bindings/kotlin-native/src/nativeMain/kotlin/org/maplibre/nativeffi/map/NorthOrientation.kt
package org.maplibre.nativeffi.map

/** North orientation used by map viewport options. */
public enum class NorthOrientation(public val nativeValue: Int) {
  UP(0),
  RIGHT(1),
  DOWN(2),
  LEFT(3),
  UNKNOWN(-1);

  public companion object {
    internal fun fromNative(nativeValue: UInt): NorthOrientation = fromNative(nativeValue.toInt())

    public fun fromNative(nativeValue: Int): NorthOrientation =
      entries.firstOrNull { it.nativeValue == nativeValue } ?: UNKNOWN
  }
}
